From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Mon, 24 Feb 2025 19:10:17 +0300
Subject: [PATCH] lithium: fast_util

This patch is based on the following mixins:
* "net/caffeinemc/mods/lithium/mixin/math/fast_util/DirectionMixin.java"
* "net/caffeinemc/mods/lithium/mixin/math/fast_util/AABBMixin.java"
By: 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0 (https://www.gnu.org/licenses/lgpl-3.0.html)

diff --git a/net/minecraft/core/AxisCycle.java b/net/minecraft/core/AxisCycle.java
index e3524e706e02683d891a8de1e552a0bd0595b0f4..438cb0f2523461b39b9b09e3d834440f17434715 100644
--- a/net/minecraft/core/AxisCycle.java
+++ b/net/minecraft/core/AxisCycle.java
@@ -35,7 +35,17 @@ public enum AxisCycle {
 
         @Override
         public Direction.Axis cycle(final Direction.Axis axis) {
-            return AXIS_VALUES[Math.floorMod(axis.ordinal() + 1, 3)];
+            // DivineMC start - lithium: fast_util
+            return switch (axis.ordinal()) {
+                case 0 -> // X
+                    Direction.Axis.Y;
+                case 1 -> // Y
+                    Direction.Axis.Z;
+                case 2 -> // Z
+                    Direction.Axis.X;
+                default -> throw new IllegalArgumentException();
+            };
+            // DivineMC end - lithium: fast_util
         }
 
         @Override
@@ -56,7 +66,17 @@ public enum AxisCycle {
 
         @Override
         public Direction.Axis cycle(final Direction.Axis axis) {
-            return AXIS_VALUES[Math.floorMod(axis.ordinal() - 1, 3)];
+            // DivineMC start - lithium: fast_util
+            return switch (axis.ordinal()) {
+                case 0 -> // X
+                    Direction.Axis.Z;
+                case 1 -> // Y
+                    Direction.Axis.X;
+                case 2 -> // Z
+                    Direction.Axis.Y;
+                default -> throw new IllegalArgumentException();
+            };
+            // DivineMC end - lithium: fast_util
         }
 
         @Override
@@ -65,6 +85,15 @@ public enum AxisCycle {
         }
     };
 
+    // DivineMC start - lithium: fast_util
+    static {
+        assert Direction.Axis.X.ordinal() == 0;
+        assert Direction.Axis.Y.ordinal() == 1;
+        assert Direction.Axis.Z.ordinal() == 2;
+        assert Direction.Axis.values().length == 3;
+    }
+    // DivineMC end - lithium: fast_util
+
     public static final Direction.Axis[] AXIS_VALUES = Direction.Axis.values();
     public static final AxisCycle[] VALUES = values();
 
diff --git a/net/minecraft/core/Direction.java b/net/minecraft/core/Direction.java
index c0fdc0a577f34159dc7d047790e26a7c0d4a0d79..c6f8856d798d0c48be93ecf42678b28c8a5910d2 100644
--- a/net/minecraft/core/Direction.java
+++ b/net/minecraft/core/Direction.java
@@ -223,7 +223,7 @@ public enum Direction implements Directional, StringRepresentable, ca.spottedlea
     }
 
     public Direction getOpposite() {
-        return this.opposite; // Paper - optimise collisions
+        return VALUES[this.oppositeIndex]; // DivineMC - lithium: fast_util
     }
 
     public Direction getClockWise(final Direction.Axis axis) {
@@ -363,7 +363,7 @@ public enum Direction implements Directional, StringRepresentable, ca.spottedlea
     }
 
     public static Direction getRandom(final RandomSource random) {
-        return Util.getRandom(VALUES, random);
+        return VALUES[random.nextInt(VALUES.length)]; // DivineMC - lithium: fast_util
     }
 
     public static Direction getApproximateNearest(final double dx, final double dy, final double dz) {
diff --git a/net/minecraft/world/phys/AABB.java b/net/minecraft/world/phys/AABB.java
index 36780d280c96f44be0e7d0dcd3467fff5e843824..42c610a37facd01676c2610a0007bc46c543432b 100644
--- a/net/minecraft/world/phys/AABB.java
+++ b/net/minecraft/world/phys/AABB.java
@@ -19,6 +19,15 @@ public class AABB {
     public final double maxY;
     public final double maxZ;
 
+    // DivineMC start - lithium: fast_util
+    static {
+        assert Direction.Axis.X.ordinal() == 0;
+        assert Direction.Axis.Y.ordinal() == 1;
+        assert Direction.Axis.Z.ordinal() == 2;
+        assert Direction.Axis.values().length == 3;
+    }
+    // DivineMC end - lithium: fast_util
+
     public AABB(final double minX, final double minY, final double minZ, final double maxX, final double maxY, final double maxZ) {
         this.minX = Math.min(minX, maxX);
         this.minY = Math.min(minY, maxY);
@@ -80,11 +89,31 @@ public class AABB {
     }
 
     public double min(final Direction.Axis axis) {
-        return axis.choose(this.minX, this.minY, this.minZ);
+        // DivineMC start - lithium: fast_util
+        return switch (axis.ordinal()) {
+            case 0 -> // X
+                this.minX;
+            case 1 -> // Y
+                this.minY;
+            case 2 -> // Z
+                this.minZ;
+            default -> throw new IllegalArgumentException();
+        };
+        // DivineMC end - lithium: fast_util
     }
 
     public double max(final Direction.Axis axis) {
-        return axis.choose(this.maxX, this.maxY, this.maxZ);
+        // DivineMC start - lithium: fast_util
+        return switch (axis.ordinal()) {
+            case 0 -> // X
+                this.maxX;
+            case 1 -> // Y
+                this.maxY;
+            case 2 -> // Z
+                this.maxZ;
+            default -> throw new IllegalArgumentException();
+        };
+        // DivineMC end - lithium: fast_util
     }
 
     @Override
